content/clause-interleaving.md

```diff
@@ -134,49 +101,44 @@ It provides the best of all worlds:
-DefTypeParamClause::=  [nl] ‘[’ DefTypeParam {‘,’ DefTypeParam} ‘]’
-DefTypeParam      ::=  {Annotation} id [HkTypeParamClause] TypeParamBounds
-DefTermParamClause::=  [nl] ‘(’ [DefTermParams] ‘)’
-UsingParamClause  ::=  [nl] ‘(’ ‘using’ (DefTermParams | FunArgTypes) ‘)’
-DefImplicitClause ::=  [nl] ‘(’ ‘implicit’ DefTermParams ‘)’
-DefTermParams     ::=  DefTermParam {‘,’ DefTermParam}
-DefTermParam      ::=  {Annotation} [‘inline’] Param
-Param             ::=  id ‘:’ ParamType [‘=’ Expr]
+DefDcl                 ::=  DefSig ‘:’ Type
+DefDef                 ::=  DefSig [‘:’ Type] ‘=’ Expr
+DefSig                 ::=  id [DefParamClauses] [DefImplicitClause]
+DefParamClauses        ::=  DefParamClauseChunk {DefParamClauseChunk}
+DefParamClauseChunk    ::=  [DefTypeParamClause] TermOrUsingParamClause {TermOrUsingParamClause}
+TermOrUsingParamClause ::=  DefTermParamClause
+                         |  UsingParamClause
+DefTypeParamClause     ::=  [nl] ‘[’ DefTypeParam {‘,’ DefTypeParam} ‘]’
+DefTypeParam           ::=  {Annotation} id [HkTypeParamClause] TypeParamBounds
+DefTermParamClause     ::=  [nl] ‘(’ [DefTermParams] ‘)’
+UsingParamClause       ::=  [nl] ‘(’ ‘using’ (DefTermParams | FunArgTypes) ‘)’
+DefImplicitClause      ::=  [nl] ‘(’ ‘implicit’ DefTermParams ‘)’
+DefTermParams          ::=  DefTermParam {‘,’ DefTermParam}
+DefTermParam           ::=  {Annotation} [‘inline’] Param
+Param                  ::=  id ‘:’ ParamType [‘=’ Expr]
 ~~~
-The main rules of interest are `DefParamClauses` and `DefParamClause`, which now allow any number of type parameter clauses, term parameter clauses and using parameter clauses, in any order.
+
+The main rules of interest are `DefParamClauses` and `DefParamClauseChunk`, which now allow any number of type parameter clauses, term parameter clauses and using parameter clauses, in any order as long as there are no two adjacent type clauses.
 
 Note that these are also used for the right-hand side of extension methods, clause interleaving thus also applies to them.
 
 It is worth pointing out that there can still only be at most one implicit parameter clause, which, if present, must be at the end.
 
 The type system and semantics naturally generalize to these new method signatures.
 
-When a method poly type consists of _n_ consecutive type parameter lists, and _m < n_ lists are specified at call site, they always correspond to the left-most _m_ type parameter lists of the signature. The right-most _(n-m)_ type parameter lists are then inferred.
-
 ### Restrictions
+
+#### Type Currying
+Type parameters cannot be curried (having two type clauses next to each other), as this would allow partial type inference, there is a big concern it would become a recommended norm to _always_ curry type parameters.
+
+Note however that, if absolutely necessary, it is still possible to curry type parameters as such: `def foo[A](using A =:= A)[B]`, since the implicit search for `A =:= A` should always succeed.
```

Comment:
You can just use `DummyImplicit` instead of `A =:= A`.

Reply:
I didn't know about that, thank you !
```suggestion
Note however that, if absolutely necessary, it is still possible to curry type parameters as such: `def foo[A](using DummyImplicit)[B]`, since the implicit search for `DummyImplicit` will always succeed.
```

Reply:
I guess the worry is that over-usage of type currying would make the code unnecessarily cryptic and more difficult to read for humans. If I'm right about that then perhaps include something similar in the text - "always" is a bold word and the universe will outlive the Scala programming language, presumably :)

Reply:
Considering the odd constructs that have come out of `fewerBraces`, this seems like a weird hill to die on. `def foo[A][B](...)` is way less cryptic than `def foo[A]: FooPartiallyApplied[A] ...`

Reply:
Well, I guess the point is that it is viewed as less cryptic than `def foo[A, B](...)`
(Anyways, I was just trying to help sharpening the text and describe in more detail the voiced concerns.)

Reply:
@morgen-peschke I appreciate your point of view
While type currying is not included in this SIP, this doesn't mean it will be banned from ever becoming a part of the language!

There are three mechanisms for partial type inference that I could come up with so far:
* type currying
* named type parameters
* allowing `?` in type application: `Vec.fill[4, ?]("Hello")` (`?` would get replaced by `String`)

I believe all of them should be carefully considered, to try to find the best way forward
(In particular, the best way forward might be to allow more than one)

To this end, a new SIP for type currying will need to be written, would you be interested in being the one to do it ?
(Of course you can take as much as you like from this SIP, in particular before commit [a492268](https://github.com/scala/improvement-proposals/pull/47/commits/a4922683dd61904aa2b0ac06fc8b8ff2ab766c20))

If I do write it instead, you will of course be welcome to comment, to try and make the best argument possible for type currying

Reply:
> I guess the worry is that over-usage of type currying would make the code unnecessarily cryptic and more difficult to read for humans. If I'm right about that then perhaps include something similar in the text - "always" is a bold word and the universe will outlive the Scala programming language, presumably :)

Thank you, I will incorporate changes to make it more clear and nuanced

> the universe will outlive the Scala programming language

That's true ^^
However, the "always" in the text is implicitly constrained by the usage of Scala (as type currying can only happen at definition), so those two statements can coexist without contradiction !

Reply:
> those two statements can coexist without contradiction

Lol. Yes different types of currying can co-exist in this ambiguous real world. (I had some extra garam masala in my spicy stew yesterday, for a change and I gave my cooking :100:  )

Reply:
@Sporarum 
> To this end, a new SIP for type currying will need to be written, would you be interested in being the one to do it ?

I'd be interested, but due to offline stuff, I wouldn't be able to pick it up until around December.

Reply:
@morgen-peschke 
Okay, if it needs to be done before, I'll do it instead, otherwise feel free to contact me when you have the time